### PR TITLE
Retry when loading a package is failed

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

Fixes: #2960 

If loading a package fails, retry it after all dependencies are downloaded.

image

For example,

1. `reboundx` depends on a shared library `librebound.so` that is in `rebound` package.
2. In the first attempt, it fails to find `librebound.so` inside the virtual file system.
3. Retry loading `reboundx` after `rebound` is downloaded, now `librebound.so` is available.

Note that 404 error happens because Emscripten loadDynamicLibrary function [tries to search the file from network](https://github.com/emscripten-core/emscripten/blob/c881fde312988ed466f0cea2a4a4d1d8e5d7403f/src/library_dylink.js#L804-L808) when it is not available on the file system.

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

### CODE REVIEW

**Comment by hoodmane** [comment]
Maybe we should throw an error in `findObject` or `findFile` as appropriate if the file isn't found to prevent the network request since we don't want it to make the request.

**Comment by ryanking13** [comment]
> Maybe we should throw an error in `findObject` or `findFile` as appropriate if the file isn't found to prevent the network request since we don't want it to make the request.

Yeah, we can try that, I think it can also help us get some more informative error messages. I'll try it in #2954 after this PR.

**Comment by ryanking13** [comment]
@hoodmane, @rth Can I ask for a review on this?

File: src/js/load-package.ts [hunk 1]
```diff
@@ -441,21 +440,46 @@ export async function loadPackage(
...
-    if (Object.keys(failed).length > 0) {
-      const failedNames = Object.keys(failed).join(", ");
+
+    if (failed.size > 0) {
+      const failedNames = Array.from(failed.keys()).join(", ");
       messageCallback(`Failed to load ${failedNames}`);
-      for (const [name, err] of Object.entries(failed)) {
+      for (const [name, err] of failed) {
         console.warn(`The following error occurred while loading ${name}:`);
         console.error(err);
```
Comment (hoodmane) [inline src/js/load-package.ts:484]:
```suggestion
        errorCallback(`The following error occurred while loading ${name}:`);
        errorCallback(err);
```

File: src/js/load-package.ts [hunk 2]
```diff
@@ -441,21 +440,46 @@ export async function loadPackage(
...
+          failed.set(name, err);
         });
     }
     await Promise.all(Object.values(packageInstallPromises));
 
+    // Retry failed packages after all dependencies are downloaded.
+    const installPromisesRetry: { [name: string]: Promise<void> } = {};
+    const toLoadRetry = new Map(
+      [...toLoadShared, ...toLoad].filter(([name, channel]) => failed.has(name))
+    );
```
Comment (hoodmane) [inline src/js/load-package.ts:452]: The other way here is a bit more efficient:
```suggestion
    const toLoadRetry = new Map(
      failed.keys().map(k => [k, toLoadShared[k] || toLoad[k]])
    );
```

File: src/js/load-package.ts [hunk 3]
```diff
@@ -441,21 +440,46 @@ export async function loadPackage(
...
+    for (const [name, channel] of toLoadRetry) {
+      console.warn(
```
Comment (hoodmane) [inline src/js/load-package.ts:454]:
```suggestion
      errorCallback(
```

File: src/js/load-package.ts [hunk 4]
```diff
@@ -441,21 +440,46 @@ export async function loadPackage(
...
         })
         .catch((err) => {
-          console.warn(err);
-          failed[name] = err;
+          failed.set(name, err);
         });
     }
     await Promise.all(Object.values(packageInstallPromises));
 
+    // Retry failed packages after all dependencies are downloaded.
```
Comment (hoodmane) [inline src/js/load-package.ts:448]:
```suggestion
    // Retry failed packages after all dependencies are downloaded: If an `.so` file 
    // from package A requires an `.so` file from package B at load time, it's possible
    // that package B is loaded before package A and fails. This is a hacky way of 
    // reordering the two dependencies (in addition to the toLoadShared, toLoad 
    // separation). Both approaches are only bandaids: What if there is a three layer
    // load time dependency?
```

**Review by hoodmane — APPROVED** [review]
I don't like this approach: this adds an extra hack (in addition to the `toLoadShared` hack) that makes the problem go away in a special case (one stage load time dependencies) without actually solving the general problem. Eventually we will see packages that still don't work (as soon as package A depends on package B depends on package C at load time this won't work).

In the meantime I suppose this is okay.

**Comment by hoodmane** [comment]
I am looking into a better solution to this.

**Comment by hoodmane** [comment]
I take it back this is a general solution if a bit hacky. I will add a year test. It also needs a better comment to explain how it works since I keep forgetting.

**Comment by rth** [comment]
Wouldn't it be better to mark `rebound` as a shared library (well maybe in a different sense than what we have now), and do not call `loadDynamicLibrary` on .so in `reboundx` until `rebound` is fully loaded? Retrying stuff for something that should be handled by dependency resolution doesn't sound optimal indeed.

It sounds like this is a similar situation with CLAPACK & scipy, or am I missing something? I did also run into related issues when playing with pyodide-pack and trying to load all .so from a single archive.

**Comment by ryanking13** [comment]
> as soon as package A depends on package B depends on package C at load time this won't work

I think this case will work, because retrying happens after downloading all .so files. But yes, this is really hacky way of solving the problem and I don't like it very much.

> Wouldn't it be better to mark rebound as a shared library (well maybe in a different sense than what we have now), and do not call loadDynamicLibrary on .so in reboundx until rebound is fully loaded? Retrying stuff for something that should be handled by dependency resolution doesn't sound optimal indeed.

We can try that, but it will add another hack: marking a python package as a shared library. I'll see if that works.

> It sounds like this is a similar situation with CLAPACK & scipy, or am I missing something? I did also run into related issues when playing with pyodide-pack and trying to load all .so from a single archive.

I don't know much about CLAPACK & scipy issue, could you explain more about it?

**Comment by ryanking13** [comment]
Aside of marking `rebound` as a shared library. I think this problem can happen between shared libraries.

If a shared library A depends on another shared library B, I think we can face this problem again.
The general solution would be downloading all dependency set first, than loading them after all downloads are complete. But I'm not sure how much it will influence the load time.

**Comment by hoodmane** [comment]
I think the approach here is sound. #3020 is a bit more satisfying but it is probably slower.

**Comment by ryanking13** [comment]
> I think the approach here is sound. #3020 is a bit more satisfying but it is probably slower.

Let's check how much #3020 will affect the load time and decide which one we should use.

**Pull request closed without merging**